src/main/java/com/playmonumenta/networkrelay/RemotePlayerManagerBungee.java

```diff
@@ -0,0 +1,185 @@
+import net.md_5.bungee.api.plugin.Listener;
+import net.md_5.bungee.event.EventHandler;
+import net.md_5.bungee.event.EventPriority;
+import org.checkerframework.checker.nullness.qual.MonotonicNonNull;
+import org.jetbrains.annotations.Nullable;
+
+public final class RemotePlayerManagerBungee extends RemotePlayerManagerAbstraction implements Listener {
+	private static @MonotonicNonNull RemotePlayerManagerBungee INSTANCE = null;
+
+	private RemotePlayerManagerBungee() {
+		INSTANCE = this;
+		try {
+			for (String shard : NetworkRelayAPI.getOnlineShardNames()) {
+				MMLog.fine("Registering shard " + shard);
+				registerServer(shard);
+			}
+		} catch (Exception ex) {
+			MMLog.severe("Failed to get remote shard names");
+			throw new RuntimeException("Failed to get remote shard names", ex);
+		}
+
+		try {
+			NetworkRelayAPI.sendExpiringBroadcastMessage(REMOTE_PLAYER_REFRESH_CHANNEL,
+				new JsonObject(),
+				REMOTE_PLAYER_MESSAGE_TTL);
+		} catch (Exception ex) {
+			MMLog.severe("Failed to broadcast to channel " + REMOTE_PLAYER_REFRESH_CHANNEL);
+		}
+	}
+
+	protected static RemotePlayerManagerBungee getInstance() {
+		if (INSTANCE == null) {
+			INSTANCE = new RemotePlayerManagerBungee();
+		}
+		return INSTANCE;
+	}
+
+	protected static String getServerId() {
+		@Nullable String shardName = null;
+		try {
+			shardName = NetworkRelayAPI.getShardName();
+		} catch (Exception e) {
+			MMLog.severe("Failed to get shard name");
+		}
+		if (shardName == null) {
+			throw new RuntimeException("Got null shard name");
+		}
+		return shardName;
+	}
+
+	protected static RemotePlayerBungee fromLocal(ProxiedPlayer player, boolean isOnline) {
+		String targetShard = ""; // TODO: add a way to get the shard name. IT IS MUCH EASIER IN VELOCITY - usb
+		return fromLocal(player, isOnline, targetShard);
+	}
+
+	protected static RemotePlayerBungee fromLocal(ProxiedPlayer player, boolean isOnline, String targetShard) {
+		return new RemotePlayerBungee(
+			getServerId(),
+			player.getUniqueId(),
+			player.getName(),
+			isOnline,
+			null,
+			targetShard
+		);
+	}
+
+	protected void refreshLocalPlayers() {
+		for (ProxiedPlayer player : ProxyServer.getInstance().getPlayers()) {
+			refreshLocalPlayer(player);
+		}
+	}
+
+	// Run this on local players whenever their information is out of date
+	protected void refreshLocalPlayer(ProxiedPlayer player) {
+		MMLog.fine("Refreshing local player " + player.getName());
+		RemotePlayerBungee localPlayer = fromLocal(player, true);
+
+		// update local player with data
+		updatePlayer(localPlayer);
+		localPlayer.broadcast();
+
+		RemotePlayerLoadedEventBungee remotePLE = new RemotePlayerLoadedEventBungee(localPlayer);
+		ProxyServer.getInstance().getPluginManager().callEvent(remotePLE);
+	}
+
+	@Nullable
+	protected RemotePlayerAbstraction remotePlayerChange(JsonObject data) {
+		RemotePlayerAbstraction player = RemotePlayerAbstraction.from(data);
+		if (player == null) {
+			// something really bad happened
```

Comment:
Well that shouldn't happen - best to put a log message here so we at least know if it does.